udp_wasip2.go

```diff
@@ -0,0 +1,135 @@
+//go:build wasip2
+
+// WASI preview 2 UDP
+
+package net
+
+import (
+	"fmt"
+	"time"
+
+	"internal/cm"
+	instancenetwork "internal/wasi/sockets/v0.2.0/instance-network"
+	"internal/wasi/sockets/v0.2.0/network"
+	"internal/wasi/sockets/v0.2.0/udp"
+	udpcreatesocket "internal/wasi/sockets/v0.2.0/udp-create-socket"
+)
+
+type wasip2UdpSocket struct {
+	udpcreatesocket.UDPSocket
+	udp.Pollable
+	*udp.IncomingDatagramStream
+	*udp.OutgoingDatagramStream
+}
+
+func createUDPSocket(af network.IPAddressFamily) (wasip2Socket, error) {
+	res := udpcreatesocket.CreateUDPSocket(af)
+	if res.IsErr() {
+		return nil, fmt.Errorf("failed to create UDP socket: %s", res.Err().String())
+	}
+
+	sock := res.OK()
+	return &wasip2UdpSocket{
+		UDPSocket: *sock,
+		Pollable:  sock.Subscribe(),
+	}, nil
+}
+
+func (s *wasip2UdpSocket) Bind(globalNetwork instancenetwork.Network, addr network.IPSocketAddress) error {
+	res := s.StartBind(globalNetwork, addr)
+	if res.IsErr() {
+		return fmt.Errorf("failed to start binding socket: %s", res.Err().String())
+	}
+
+	res = s.FinishBind()
+	if res.IsErr() {
+		return fmt.Errorf("failed to finish binding socket: %s", res.Err().String())
+	}
+
+	return nil
+}
+
+func (s *wasip2UdpSocket) Listen(backlog int) error {
+	fmt.Println("wasip2 UDP listen TODO:", backlog) ///
+	return nil
+}
+
+func (s *wasip2UdpSocket) Accept() (wasip2Socket, *network.IPSocketAddress, error) {
+	return nil, nil, fmt.Errorf("wasip2 UDP sockets do not support Accept")
+}
+
+func (s *wasip2UdpSocket) Connect(globalNetwork instancenetwork.Network, host string, ip network.IPSocketAddress) error {
+	res := s.UDPSocket.Stream(cm.Some(ip))
+
+	if res.IsErr() {
+		return fmt.Errorf("failed to connect UDP socket: %s", res.Err().String())
+	}
+
+	s.IncomingDatagramStream, s.OutgoingDatagramStream = &res.OK().F0, &res.OK().F1
+
+	return nil
+}
+
+func (c wasip2UdpSocket) Send(buf []byte, flags int, deadline time.Time) (int, error) {
+	if flags != 0 {
+		fmt.Println("wasip2 UDP send flags TODO:", flags) ///
+	}
+
+	if c.OutgoingDatagramStream == nil {
+		return -1, fmt.Errorf("send called on a socket without open streams")
+	}
+
+	res := c.OutgoingDatagramStream.CheckSend()
+	if res.IsErr() {
+		return -1, fmt.Errorf("failed to write to output stream: %s", res.Err().String())
+	}
+
+	if *res.OK() == 0 {
+		c.Block()
+	}
+
+	a := []udp.OutgoingDatagram{
+		{
+			Data: cm.NewList(&buf[0], len(buf)),
+			// RemoteAddress: cm.Some(c.raddr),
```

Comment:
Should this be commented out?  If so, remove.  Otherwise, probably worth fixing.  Does the remote address get set somewhere else?